README.md
# Vertical Style Card
Vertical Style is a custom Lovelace card for Home Assistant who let you group diferent cards in one vertical-stack card, with diferent style options, like title, card color, to have a border or not. This is very useful for cards like custom:button-card or picture-entity, giving the option for add a title and grouping in a column.

![Vertical Style Examples](example.PNG)


## Options

| Name | Type | Default | Description
| ---- | ---- | ------- | -----------
| type | string | **Required** | `custom:vertical-style-card`
| cards | list | **Required** | List of cards
| title | string | **Optional** | Card title
| style | object | **Optional** | Options: font_size, border,background_color


## Style Object

| Name | Type | Default | Options
| ---- | ---- | ------- | -----------
| font_size | string | **Optional** | `Example: 20px`
| borderRadius | string | 20px | `Example: 2px`
| border | boolean | **true** | Show or not the card border
| background_color | string | **Optional** | Options: rgb, var(--), hex
| firstShadowAlpha | string | 0.14
| secondShadowAlpha | string | 0.12
| thirdShadowAlpha | string | 0.15
| padding | string | 5px



## Installation

1. Install the `vertical-style-card` card by copying `vertical-style-card.js` to `<config directory>/www/vertical-style-card.js`

Bash:
```bash
wget https://raw.githubusercontent.com/matisaul/vertical-style-card/master/vertical-style-card.js
mv vertical-style-card.js /config/www/
```

2. Link `vertical-style-card` inside you `ui-lovelace.yaml` 

```yaml
resources:
  - url: /local/vertical-style-card.js?v=0.1
    type: js
```

3. Add a custom card in your `ui-lovelace.yaml`




#### To add vertical-style-card to custom_updater:

```yaml
custom_updater:
  card_urls: 
    - https://raw.githubusercontent.com/matisaul/vertical-style-card/master/tracker.json
```


**Example**

```yaml
- type: custom:vertical-style-card
  title: Scenes
  style:
    font_size: 18px
    border: false
    background_color: var(--secondary-background-color)
  cards:
    - type: horizontal-stack
      cards:
        - type: picture-entity
          image: /local/images/love.png
          entity: scene.romantic
          name: Romantic
          show_state: false
          tap_action: toggle
        - type: picture-entity
          image: /local/images/soccer.png
          entity: scene.soccer
          name: Soccer
          show_state: false
          tap_action: toggle 
    - type: horizontal-stack
      cards:
        - type: picture-entity
          image: /local/images/party.png
          entity: scene.party
          name: Party
          show_state: false
          tap_action: toggle
        - type: picture-entity
          image: /local/images/netflix.png
          entity: scene.netflix
          name: Netflix
          show_state: false
          tap_action: toggle

 
```

## Credits
- [ofekashery](https://github.com/ofekashery)
- [ciotlosm](https://github.com/ciotlosm)
- [thomasloven](https://github.com/thomasloven)
- 



changelog.md
# 0.1.4

Styles fix.
# 0.1.3

Styles fix.

## 0.1.2

* An "element.setConfig is not a function" error has been fixed.
* Border radius
  * Default value changed from 2px to 20px.
  * `borderRadius` parameter added into style config section.
* Parameters `firstShadowAlpha`, `secondShadowAlpha` and `thirdShadowAlpha` have been added into style config section for fine tune shadows.
* A `padding` parameter has been added into style config section.
## 0.1
Fixed problem with HA > 0.80.0

## 0.0.2
Add posibility to don't add the style option

## 0.0.1
Initial release


vertical-style-card.js
class VerticalStyleCard extends HTMLElement {
    constructor() {
        super();
        // Make use of shadowRoot to avoid conflicts when reusing
        this.attachShadow({ mode: 'open' });
    }
    setConfig(config) {
        if (!config || !config.cards || !Array.isArray(config.cards)) {
            throw new Error('Card config incorrect');
        }
        const cardConfig = Object.assign({}, config);
		cardConfig.style = Object.assign({}, config.style);
	
		var prop = {  "border": null  , "background_color":null , "font_size":null}  ;
		
		var cardSize = Object.keys(cardConfig.style);
		
		if (cardSize.length == 0) Object.assign(cardConfig.style, prop);

		if (cardConfig.style.border === undefined || cardConfig.style.border == null ) cardConfig.style.border = true;
		if (!cardConfig.style.background_color || cardConfig.style.background_color == null ) cardConfig.style.background_color = 'var(--paper-card-background-color)';
		if (!cardConfig.style.font_size || cardConfig.style.font_size == null ) cardConfig.style.font_size = 'var(--paper-font-headline_-_font-size)';
        if (!cardConfig.style.borderRadius || cardConfig.style.borderRadius == null ) cardConfig.style.borderRadius = '20px';
        if (!cardConfig.style.firstShadowAlpha || cardConfig.style.firstShadowAlpha == null ) cardConfig.style.firstShadowAlpha = '0.14';
        if (!cardConfig.style.secondShadowAlpha || cardConfig.style.secondShadowAlpha == null ) cardConfig.style.secondShadowAlpha = '0.12';
        if (!cardConfig.style.thirdShadowAlpha || cardConfig.style.thirdShadowAlpha == null ) cardConfig.style.thirdShadowAlpha = '0.15';
        if (!cardConfig.style.padding || cardConfig.style.padding == null ) cardConfig.style.padding = '5px';

		this.style.background = cardConfig.style.background_color;
		this.style.padding = cardConfig.style.padding + ' ' + cardConfig.style.padding + ' ' + cardConfig.style.padding + ' ' + cardConfig.style.padding;

		if (cardConfig.style.border == true){
			this.style.boxShadow = '0 2px 2px 0 rgba(0, 0, 0, ' + cardConfig.style.firstShadowAlpha + '), 0 1px 5px 0 rgba(0, 0, 0, ' + cardConfig.style.secondShadowAlpha + '), 0 3px 1px -2px rgba(0, 0, 0, ' + cardConfig.style.thirdShadowAlpha + ')';
			this.style.borderRadius = cardConfig.style.borderRadius;
		}

        const root = this.shadowRoot;
        while (root.lastChild) {
            root.removeChild(root.lastChild);
        }

        this._refCards = [];
        if (config.title) {
            const title = document.createElement("div");
            title.className = "header";
            title.style = "font-family: var(--paper-font-headline_-_font-family); -webkit-font-smoothing: var(--paper-font-headline_-_-webkit-font-smoothing); font-size: " + cardConfig.style.font_size + "; font-weight: var(--paper-font-headline_-_font-weight); letter-spacing: var(--paper-font-headline_-_letter-spacing); line-height: var(--paper-font-headline_-_line-height);text-rendering: var(--paper-font-common-expensive-kerning_-_text-rendering);opacity: var(--dark-primary-opacity);padding: 5px 16px 10px 16px";
            title.innerHTML = '<div class="name">' + config.title + '</div>';
            root.appendChild(title);
        }
        let element;
        config.cards.forEach(item => {
            if (item.type.startsWith("custom:")){
                element = document.createElement(`${item.type.substr("custom:".length)}`);
            } else {
                element = document.createElement(`hui-${item.type}-card`);
            }
            if (element.setConfig != undefined)
                element.setConfig(item);
            root.appendChild(element);
            this._refCards.push(element);
        });
    }

    set hass(hass) {
        if (this._refCards) {
            this._refCards.forEach((card) => {
                card.hass = hass;
            });
        }
    }

    connectedCallback() {
        this._refCards.forEach((element) => {
            let fn = () => {
                this._card(element);
            };

            if(element.updateComplete) {
                element.updateComplete.then(fn);
            } else {
                fn();
            }
        });
    }

    _card(element) {
        if (element.shadowRoot) {
            if (!element.shadowRoot.querySelector('ha-card')) {
                let searchEles = element.shadowRoot.getElementById("root");
                if (!searchEles) {
                    searchEles = element.shadowRoot.getElementById("card");
                }
                if (!searchEles) return;
                searchEles = searchEles.childNodes;
                for (let i = 0; i < searchEles.length; i++) {
                    if(searchEles[i].style !== undefined){
                        searchEles[i].style.margin = "2px";
                    }
                    this._card(searchEles[i]);
                }
            } else {
                element.shadowRoot.querySelector('ha-card').style.boxShadow = 'none';
            }
        } else {
            if (typeof element.querySelector === 'function' && element.querySelector('ha-card')) {
                element.querySelector('ha-card').style.boxShadow = 'none';
            }
            let searchEles = element.childNodes;
            for (let i = 0; i < searchEles.length; i++) {
                if (searchEles[i] && searchEles[i].style) {
                    searchEles[i].style.margin = "2px";
                }
                this._card(searchEles[i]);
            }
        }
    }
    
    getCardSize() {
        let totalSize = 0;
        this._refCards.forEach((element) => {
            totalSize += typeof element.getCardSize === 'function' ? element.getCardSize() : 1;
        });
        return totalSize;
    }
}

customElements.define('vertical-style-card', VerticalStyleCard);
